README.md
# 2019 — BERT Phishing Email Classifier

Fine-tune BERT-base-uncased for binary classification of phishing vs. benign emails.

## Approach

1. Tokenize emails with BERT's WordPiece tokenizer (`bert-base-uncased`)
2. Add a single linear classification head on top of `[CLS]`
3. Fine-tune end-to-end on labeled phishing dataset
4. Evaluate accuracy, precision/recall, F1

## Files

- `src/bert_classifier.py` — BertForSequenceClassification fine-tune loop
- `src/data.py` — dataset loader (placeholder — drop your CSV at `data/emails.csv`)

## Run

```bash
pip install -r requirements.txt
python src/bert_classifier.py --train data/emails.csv --epochs 3
```

## Dataset

Use Nazario Phishing Corpus (public, classic) or any labeled email CSV with columns
`text` and `label` (0 = benign, 1 = phishing). Don't commit real emails — this is for
local study only.



CHANGELOG.md
# Changelog

## [0.3.0] — 2019-12-22
- Dataset documentation
- Stratified split + `drop_empty` for imbalanced corpora
- Pin transformers >= 2.5 (Trainer API arrives mid-2020)

## [0.2.0] — 2019-09-25
- Inference CLI (predict.py)
- Evaluation script with classification report + ROC + CM plots
- Email normalization helpers (URL/IP/email tokenization)

## [0.1.0] — 2019-08-22
- BertForSequenceClassification fine-tune loop
- EmailDataset with WordPiece tokenization


tests/test_data.py
"""Tests for src.data — EmailDataset + stratified split."""
import csv
import os
import tempfile
import unittest

from data import EmailDataset, stratified_split


class FakeTokenizer:
    """Minimal tokenizer that mimics BertTokenizer's encode_plus return shape."""
    def encode_plus(self, text, **kwargs):
        max_len = kwargs.get('max_length', 16)
        ids = [101] + [hash(w) % 30000 for w in text.split()][:max_len - 2] + [102]
        ids += [0] * (max_len - len(ids))
        mask = [1 if i != 0 else 0 for i in ids]
        return {
            'input_ids': [ids],
            'attention_mask': [mask],
        }


def _write_csv(rows, path):
    with open(path, 'w', newline='') as f:
        w = csv.DictWriter(f, fieldnames=['text', 'label'])
        w.writeheader()
        for r in rows:
            w.writerow(r)


class TestEmailDataset(unittest.TestCase):
    def test_basic_load(self):
        with tempfile.TemporaryDirectory() as d:
            path = os.path.join(d, 'a.csv')
            _write_csv([
                {'text': 'click here now', 'label': '1'},
                {'text': 'project update', 'label': '0'},
            ], path)
            ds = EmailDataset(path, FakeTokenizer(), max_len=16)
            self.assertEqual(len(ds), 2)
            item = ds[0]
            self.assertIn('input_ids', item)
            self.assertIn('attention_mask', item)
            self.assertIn('label', item)

    def test_skips_empty_text(self):
        with tempfile.TemporaryDirectory() as d:
            path = os.path.join(d, 'a.csv')
            _write_csv([
                {'text': '', 'label': '1'},
                {'text': 'hi', 'label': '0'},
            ], path)
            ds = EmailDataset(path, FakeTokenizer(), max_len=16)
            self.assertEqual(len(ds), 1)


class TestStratifiedSplit(unittest.TestCase):
    def test_proportions_preserved(self):
        labels = [0]*80 + [1]*20
        train, val = stratified_split(labels, val_frac=0.2, seed=0)
        # Should preserve 4:1 ratio in both partitions
        train_l = [labels[i] for i in train]
        val_l = [labels[i] for i in val]
        self.assertAlmostEqual(train_l.count(1) / len(train_l), 0.2, places=1)
        self.assertAlmostEqual(val_l.count(1) / len(val_l), 0.2, places=1)


if __name__ == '__main__':
    unittest.main()


tests/test_utils.py
"""Tests for src.utils — text neutralization + URL feature extraction."""
import unittest

from src.utils import (
    URL_RE,
    IP_RE,
    EMAIL_ADDR_RE,
    decode_subject,
    neutralize,
    url_features,
    normalize_email,
)


class TestRegexes(unittest.TestCase):
    def test_url_regex_matches_http_and_https(self):
        self.assertTrue(URL_RE.search("see http://x.com/a"))
        self.assertTrue(URL_RE.search("see HTTPS://X.COM/path?q=1"))

    def test_url_regex_does_not_match_bare_words(self):
        self.assertFalse(URL_RE.search("just words here, no url"))

    def test_ip_regex(self):
        self.assertTrue(IP_RE.search("from 192.168.1.1 today"))
        self.assertFalse(IP_RE.search("v1.2.3 release"))

    def test_email_regex(self):
        self.assertTrue(EMAIL_ADDR_RE.search("contact ops+abuse@example.com"))


class TestNeutralize(unittest.TestCase):
    def test_replaces_urls(self):
        out = neutralize("click http://bad.example/path or http://other.example")
        self.assertNotIn("bad.example", out)
        self.assertIn("<URL>", out)

    def test_replaces_ips(self):
        out = neutralize("from 10.0.0.1 came a message")
        self.assertIn("<IP>", out)

    def test_collapses_whitespace(self):
        out = neutralize("hello    world\n\nhi")
        self.assertEqual(out, "hello world hi")


class TestURLFeatures(unittest.TestCase):
    def test_no_urls(self):
        f = url_features("plain text")
        self.assertEqual(f['n_urls'], 0)
        self.assertFalse(f['has_ip_url'])

    def test_ip_url_detected(self):
        f = url_features("http://192.168.0.1/login")
        self.assertEqual(f['n_urls'], 1)
        self.assertTrue(f['has_ip_url'])

    def test_mixed_protocol(self):
        f = url_features("http://a.com and https://b.com")
        self.assertTrue(f['mixed_protocol'])


class TestDecodeSubject(unittest.TestCase):
    def test_plain(self):
        self.assertEqual(decode_subject("Hello"), "Hello")

    def test_quoted_printable(self):
        # =?utf-8?Q?H=C3=A9llo?= → "Héllo"
        self.assertEqual(decode_subject("=?utf-8?Q?H=C3=A9llo?="), "Héllo")


class TestNormalizeEmail(unittest.TestCase):
    def test_strips_headers_and_neutralizes(self):
        raw = (
            "From: bad@evil.example\nSubject: Click\n\n"
            "Hello, click http://phish.example/login now from 1.2.3.4."
        )
        out = normalize_email(raw)
        self.assertIn("<URL>", out)
        self.assertIn("<IP>", out)
        self.assertNotIn("From:", out)


if __name__ == '__main__':
    unittest.main()


src/data.py
"""Email dataset loader. Expects CSV with `text` and `label` columns."""
from collections import defaultdict

import pandas as pd
import torch
from torch.utils.data import Dataset


class EmailDataset(Dataset):
    def __init__(self, csv_path, tokenizer, max_len=256, drop_empty=True):
        df = pd.read_csv(csv_path)
        if drop_empty:
            df = df[df['text'].astype(str).str.strip().str.len() > 0]
        self.texts = df['text'].astype(str).tolist()
        self.labels = df['label'].astype(int).tolist()
        self.tokenizer = tokenizer
        self.max_len = max_len

    def __len__(self):
        return len(self.texts)

    def __getitem__(self, idx):
        enc = self.tokenizer.encode_plus(
            self.texts[idx],
            max_length=self.max_len,
            pad_to_max_length=True,  # the 2019 kwarg
            truncation=True,
            return_tensors='pt',
        )
        ids = enc['input_ids']
        mask = enc['attention_mask']
        if not isinstance(ids, torch.Tensor):
            ids = torch.tensor(ids, dtype=torch.long)
            mask = torch.tensor(mask, dtype=torch.long)
        return {
            'input_ids': ids.squeeze(0) if ids.dim() == 2 else ids,
            'attention_mask': mask.squeeze(0) if mask.dim() == 2 else mask,
            'label': torch.tensor(self.labels[idx], dtype=torch.long),
        }


def stratified_split(labels, val_frac=0.1, seed=0):
    """Return (train_indices, val_indices) preserving label proportions.

    Useful when classes are imbalanced; a random split can bury a small class
    entirely in the train set.
    """
    import random
    rng = random.Random(seed)
    by_label = defaultdict(list)
    for i, y in enumerate(labels):
        by_label[int(y)].append(i)
    train_idx, val_idx = [], []
    for label, idxs in by_label.items():
        rng.shuffle(idxs)
        n_val = max(1, int(round(len(idxs) * val_frac)))
        val_idx.extend(idxs[:n_val])
        train_idx.extend(idxs[n_val:])
    rng.shuffle(train_idx)
    rng.shuffle(val_idx)
    return train_idx, val_idx


src/utils.py
"""Text preprocessing utilities for the phishing classifier.

Email-specific normalization: strip headers, decode quoted-printable,
neutralize URLs (replace with `<URL>`) and IPs (replace with `<IP>`).
The goal is to keep the structural signal (e.g. "click here", URL pattern,
mismatched display vs. href) while not letting the model overfit to
specific URL strings in the training set.
"""
import re
from email import message_from_string
from email.header import decode_header


URL_RE = re.compile(r'https?://[^\s<>"\'`]+', re.IGNORECASE)
IP_RE = re.compile(r'\b(?:\d{1,3}\.){3}\d{1,3}\b')
EMAIL_ADDR_RE = re.compile(r'[\w\.\+-]+@[\w\.-]+\.[a-zA-Z]{2,}')
WS_RE = re.compile(r'\s+')


def strip_headers(raw: str) -> str:
    """Return only the message body from an RFC 822 string."""
    try:
        msg = message_from_string(raw)
    except Exception:
        return raw
    if msg.is_multipart():
        parts = []
        for part in msg.walk():
            ctype = part.get_content_type()
            if ctype.startswith('text/'):
                payload = part.get_payload(decode=True)
                if payload:
                    parts.append(_safe_decode(payload))
        return '\n'.join(parts)
    payload = msg.get_payload(decode=True)
    if payload is None:
        return msg.get_payload() or ''
    return _safe_decode(payload)


def _safe_decode(raw: bytes) -> str:
    for enc in ('utf-8', 'latin-1', 'cp1252'):
        try:
            return raw.decode(enc)
        except UnicodeDecodeError:
            continue
    return raw.decode('utf-8', errors='replace')


def decode_subject(s: str) -> str:
    """Decode RFC 2047 encoded-word headers (=?utf-8?Q?...?=)."""
    parts = decode_header(s)
    out = []
    for txt, enc in parts:
        if isinstance(txt, bytes):
            out.append(txt.decode(enc or 'utf-8', errors='replace'))
        else:
            out.append(txt)
    return ''.join(out)


def neutralize(text: str, replace_urls=True, replace_ips=True,
               replace_emails=True) -> str:
    """Replace URLs/IPs/email addresses with placeholder tokens."""
    if replace_urls:
        text = URL_RE.sub(' <URL> ', text)
    if replace_ips:
        text = IP_RE.sub(' <IP> ', text)
    if replace_emails:
        text = EMAIL_ADDR_RE.sub(' <EMAIL> ', text)
    return WS_RE.sub(' ', text).strip()


def url_features(text: str) -> dict:
    """Surface-level URL-based features that a classifier could attend to."""
    urls = URL_RE.findall(text)
    if not urls:
        return {'n_urls': 0, 'has_ip_url': False, 'max_url_len': 0,
                'mixed_protocol': False}
    has_ip = any(IP_RE.search(u) for u in urls)
    protocols = set(u.split(':', 1)[0].lower() for u in urls)
    return {
        'n_urls': len(urls),
        'has_ip_url': has_ip,
        'max_url_len': max(len(u) for u in urls),
        'mixed_protocol': len(protocols) > 1,
    }


def normalize_email(raw: str) -> str:
    """Full pipeline: headers → body → neutralize → collapse whitespace."""
    body = strip_headers(raw)
    return neutralize(body)


src/predict.py
"""CLI for running inference with a fine-tuned phishing classifier.

Usage:
    python src/predict.py --model-dir checkpoints/bert-phishing/ --text "..."
    python src/predict.py --model-dir checkpoints/bert-phishing/ --csv emails.csv
"""
import argparse
import csv
import sys

import torch
from transformers import BertTokenizer, BertForSequenceClassification


def load_model(model_dir, device):
    tokenizer = BertTokenizer.from_pretrained(model_dir)
    model = BertForSequenceClassification.from_pretrained(model_dir).to(device)
    model.eval()
    return tokenizer, model


def predict_one(tokenizer, model, text, device, max_len=256):
    enc = tokenizer.encode_plus(
        text,
        add_special_tokens=True,
        max_length=max_len,
        truncation=True,
        padding='max_length',
        return_tensors='pt',
    )
    input_ids = enc['input_ids'].to(device)
    attention_mask = enc['attention_mask'].to(device)
    with torch.no_grad():
        out = model(input_ids, attention_mask=attention_mask)
    logits = out[0] if isinstance(out, tuple) else out.logits
    probs = torch.softmax(logits, dim=-1)[0]
    pred = int(torch.argmax(probs).item())
    return pred, probs.cpu().tolist()


def main():
    p = argparse.ArgumentParser()
    p.add_argument('--model-dir', required=True)
    p.add_argument('--text', help='single email text to classify')
    p.add_argument('--csv', help='CSV with `text` column')
    p.add_argument('--threshold', type=float, default=0.5,
                   help='probability threshold for phishing class')
    args = p.parse_args()

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    tokenizer, model = load_model(args.model_dir, device)

    if args.text:
        pred, probs = predict_one(tokenizer, model, args.text, device)
        label = 'phishing' if probs[1] >= args.threshold else 'benign'
        print(f"{label}\tphishing_prob={probs[1]:.4f}")
        return

    if args.csv:
        w = csv.writer(sys.stdout)
        w.writerow(['idx', 'pred', 'phishing_prob'])
        with open(args.csv) as f:
            for i, row in enumerate(csv.DictReader(f)):
                pred, probs = predict_one(tokenizer, model, row['text'], device)
                label = 'phishing' if probs[1] >= args.threshold else 'benign'
                w.writerow([i, label, f"{probs[1]:.4f}"])
        return

    print('error: provide --text or --csv', file=sys.stderr)
    sys.exit(2)


if __name__ == '__main__':
    main()


src/evaluate.py
"""Evaluation script for the phishing classifier.

Computes classification report (precision/recall/F1 per class), confusion
matrix, and ROC curve. Writes a JSON summary + PNG plots to results/.
"""
import argparse
import json
import os

import numpy as np
import torch
from torch.utils.data import DataLoader
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    roc_curve,
    auc,
)
from transformers import BertTokenizer, BertForSequenceClassification

from data import EmailDataset


def collect_predictions(model, tokenizer, csv_path, device, batch_size=32, max_len=256):
    ds = EmailDataset(csv_path, tokenizer, max_len=max_len)
    loader = DataLoader(ds, batch_size=batch_size)
    all_probs = []
    all_labels = []
    model.eval()
    with torch.no_grad():
        for batch in loader:
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['label'].to(device)
            out = model(input_ids, attention_mask=attention_mask)
            logits = out[0] if isinstance(out, tuple) else out.logits
            probs = torch.softmax(logits, dim=-1)
            all_probs.append(probs.cpu().numpy())
            all_labels.append(labels.cpu().numpy())
    return np.vstack(all_probs), np.concatenate(all_labels)


def maybe_plot(probs, labels, outdir):
    try:
        import matplotlib
        matplotlib.use('Agg')
        import matplotlib.pyplot as plt
    except ImportError:
        return None

    fpr, tpr, _ = roc_curve(labels, probs[:, 1])
    roc_auc = auc(fpr, tpr)
    fig, ax = plt.subplots()
    ax.plot(fpr, tpr, label=f'AUC = {roc_auc:.3f}')
    ax.plot([0, 1], [0, 1], 'k--', alpha=0.4)
    ax.set_xlabel('False Positive Rate')
    ax.set_ylabel('True Positive Rate')
    ax.legend()
    ax.set_title('Phishing classifier — ROC')
    fig.tight_layout()
    fig.savefig(os.path.join(outdir, 'roc.png'))
    plt.close(fig)

    cm = confusion_matrix(labels, probs.argmax(axis=1))
    fig, ax = plt.subplots()
    im = ax.imshow(cm, cmap='Blues')
    for i in range(2):
        for j in range(2):
            ax.text(j, i, str(cm[i, j]), ha='center', va='center')
    ax.set_xticks([0, 1]); ax.set_yticks([0, 1])
    ax.set_xticklabels(['benign', 'phishing'])
    ax.set_yticklabels(['benign', 'phishing'])
    ax.set_xlabel('Predicted'); ax.set_ylabel('True')
    ax.set_title('Phishing classifier — Confusion matrix')
    fig.colorbar(im, ax=ax)
    fig.tight_layout()
    fig.savefig(os.path.join(outdir, 'confusion_matrix.png'))
    plt.close(fig)
    return roc_auc


def main():
    p = argparse.ArgumentParser()
    p.add_argument('--model-dir', required=True)
    p.add_argument('--test-csv', required=True)
    p.add_argument('--out', default='results/eval_summary.json')
    args = p.parse_args()

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    tokenizer = BertTokenizer.from_pretrained(args.model_dir)
    model = BertForSequenceClassification.from_pretrained(args.model_dir).to(device)

    probs, labels = collect_predictions(model, tokenizer, args.test_csv, device)

    preds = probs.argmax(axis=1)
    report = classification_report(labels, preds,
                                   target_names=['benign', 'phishing'],
                                   output_dict=True)

    os.makedirs(os.path.dirname(args.out), exist_ok=True)
    auc_val = maybe_plot(probs, labels, os.path.dirname(args.out))

    summary = {
        'classification_report': report,
        'auc': auc_val,
        'n_examples': int(len(labels)),
    }
    with open(args.out, 'w') as f:
        json.dump(summary, f, indent=2)
    print(json.dumps(summary, indent=2))


if __name__ == '__main__':
    main()


src/bert_classifier.py
"""BERT fine-tune for phishing classification. transformers 2.x style."""
import argparse

import torch
from torch.utils.data import DataLoader, random_split
from transformers import (
    BertTokenizer,
    BertForSequenceClassification,
    AdamW,
    get_linear_schedule_with_warmup,
)
from sklearn.metrics import classification_report

from data import EmailDataset


def train(csv_path, epochs=3, batch_size=16, lr=2e-5, max_len=256):
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
    model = BertForSequenceClassification.from_pretrained(
        'bert-base-uncased', num_labels=2
    ).to(device)

    ds = EmailDataset(csv_path, tokenizer, max_len=max_len)
    n_val = max(1, int(0.1 * len(ds)))
    train_ds, val_ds = random_split(ds, [len(ds) - n_val, n_val])
    train_loader = DataLoader(train_ds, batch_size=batch_size, shuffle=True)
    val_loader = DataLoader(val_ds, batch_size=batch_size)

    total_steps = len(train_loader) * epochs
    optim = AdamW(model.parameters(), lr=lr, correct_bias=False)
    sched = get_linear_schedule_with_warmup(optim, num_warmup_steps=0,
                                            num_training_steps=total_steps)

    for ep in range(epochs):
        model.train()
        for batch in train_loader:
            ids = batch['input_ids'].to(device)
            mask = batch['attention_mask'].to(device)
            labels = batch['label'].to(device)
            outputs = model(ids, attention_mask=mask, labels=labels)
            loss = outputs[0]
            optim.zero_grad()
            loss.backward()
            torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
            optim.step()
            sched.step()

        model.eval()
        ys, ps = [], []
        with torch.no_grad():
            for batch in val_loader:
                ids = batch['input_ids'].to(device)
                mask = batch['attention_mask'].to(device)
                logits = model(ids, attention_mask=mask)[0]
                ps += logits.argmax(-1).cpu().tolist()
                ys += batch['label'].tolist()
        print('=== epoch {} ==='.format(ep + 1))
        print(classification_report(ys, ps, target_names=['benign', 'phishing']))

    model.save_pretrained('./bert_phishing')
    tokenizer.save_pretrained('./bert_phishing')
    print('Saved to ./bert_phishing')


if __name__ == '__main__':
    p = argparse.ArgumentParser()
    p.add_argument('--train', required=True, help='Path to training CSV')
    p.add_argument('--epochs', type=int, default=3)
    args = p.parse_args()
    train(args.train, epochs=args.epochs)
